src/Data/Str.php

```diff
@@ -169,4 +169,31 @@ public static function ConvertCase(string $str, $encoding = null)
         // returns original string when mbstring is not active.
         return $str;
     }
+
+    /**
+     * Checks if given string is a valid Base64 encoded string.
+     *
+     * @param string $str String to be tested
+     *
+     * @return bool wether it is a valid base64 encoded string or not
+     */
+    public static function isBase64(string $str): bool
+    {
+        return $str === base64_encode(base64_decode($str));
+    }
+
+    /**
+     * Return only a portion of given string.
+     *
+     * @param string   $str      input string to process
+     * @param int      $start    where to start the cut
+     * @param int|null $length   how many characters to return
+     * @param string   $encoding optional encoding to us
+     *
+     * @return string
+     */
+    public static function substring(string $str, int $start, int $length = null, $encoding = null): string
```

Comment:
IMHO, the `substring` method should be changed as follows:

```php
public static function substring(string $str, int $start, ?int $length, ?$encoding): string
```

That means the `$length` and `$encoding` will be possible to be `null`.

And this syntax is available since `php-7.1`.

Reply:
with this change one would need to explicitly set encoding to null instead of omitting it, else:

ArgumentCountError: Too few arguments to function Zest\Data\Str::substring(), 3 passed in /c/workspace/hacktoberfest/Zest_Framework/Tests/Data/StrTest.php on line 62 and exactly 4 expected

Reply:
Hi @2mt-heuser, sorry. My suggested code is incorrect. Please try this:

```
public static function substring(string $str, int $start, ?int $length, ?string $encoding): string
```

Reply:
@peter279k already did. but the question mark only allows null or string aka making the parameter required instead of optional.
Without a default value it will crash if you just call the function with 3 parameters.

```php
public static function substring(string $str, int $start, ?int $length, ?string $encoding = null)
```
this would work, but in thise case ?string wouldn't make any sense as we're just setting it to null anyway, so we could just leave it without the ?
```php
public static function substring(string $str, int $start, ?int $length, string $encoding = null)
```

Same thing applies to ?int $length. If we only call substring with 2 parameters it should use default encoding and return the string from $start until the end but it will crash with ?int $length and not setting it to null
```php
$this->assertSame('def', Str::substring('abcdef', 3));
```
So best bet would be to stick to the current version, right?!
Sounds okay to you?

Reply:
Going ahead with this.